fix: Fix innerHTML get to not escape text if a decendant of a certain element types

craigpetchell opened the pull request:
Text nodes with particular ancestors (e.g. script, and style) should be treated as raw text when returning the value of innerHTML.

See section 8.4 of https://www.w3.org/TR/2008/WD-html5-20080610/serializing.html#html-fragment

Commit:
fix: Fix innerHTML get to not escape text if a decendant of a particular set of elements as per the

craigpetchell commented on src/index.js:
```diff
@@ -609,12 +610,39 @@ const members = {
+      // Text nodes with these ancestors should be treated as raw text
+      // See section 8.4 of
+      // https://www.w3.org/TR/2008/WD-html5-20080610/serializing.html#html-fragment
+      const rawTextNodeAncestors = new Set([
```
Will change this to an object instead, as Set is not support in Opera, and I don't want to introduce another dependency just for that.

Commit:
fix: Cross browser compatibility

adevnadia commented on src/index.js:
```diff
@@ -609,12 +610,40 @@ const members = {
+          while (parentNode) {
+            if (parentNode.nodeType === 1 && // only for elements
+                parentNode.nodeName.toLowerCase() in rawTextNodeAncestors) {
+              getOuterHtml = getRawTextContent;
```
If a text node is inside a normal Element node, with this code you will traverse up until the root node. Is it really necessary? You have tests that test only one simple case of a text node inside a 'script' element.

craigpetchell replied:
According to the w3c spec (linked above) it is, though for the use case I was fixing (script tag), it is probably unnecessary.

> If one of the **ancestors** of current node is a ...

adevnadia replied:
Still feels like it's a horrible overkill. @treshugart what do you think?

treshugart replied:
I think it'd be good to be W3C spec compliant if it's not too much work or performance overhead. It looks like this would kill perf with large numbers of children because `children * levelsDeep`. Would it work the same if you instead detected the context based on the root node that `innerHTML` was called in instead of the children?

craigpetchell replied:
In terms of the tags where this behaviour is described:
- `style` - text is the immediate child
- `script` - text is the immediate child
- `iframe` - useful content is attributes (src, etc)
- **`noscript`** - can contain HTML
- `xmp` - deprecated
- `noembed` - deprecated
- `noframes` - deprecated
- `plaintext` - deprecated

The only tag that is really impacted is the `noscript` tag, which if we don't look up the tree of a text node, we will encode the innerHTML incorrectly. Perhaps this is something we can live with, as it's probably unlikely to get used too often.

treshugart replied:
For all of those cases can't you cache whether or not the current node is one of those or is a descendant of one of those types and then do stuff with the children based on that value instead of checking within the loop?

craigpetchell replied:
The problem with caching is knowing when to expire (e.g. moving an element around the dom). I'd lean toward not looking up the tree, and have noscript potentially broken. Not sure how often innerHTML will be used on a noscript element, if used at all.

craigpetchell replied:
So. It turns out Chrome escapes <noscript> contents anyway. FF does not. Both both do not represent any tags inside a <noscript> tag as elements anyway.
So the need to walk up the tree seems unnecessary anyway.
So as much as I built the optimisation before checking, it's not necessary. :/

Going to just optimise the check the current node, and any element nodes I iterate over.

treshugart commented on src/index.js:
```diff
@@ -609,12 +610,40 @@ const members = {
+        if (node.nodeType === 3) {
+          let parentNode = node.parentNode;
+          while (parentNode) {
+            if (parentNode.nodeType === 1 && // only for elements
```
You can self-document this check by using `parentNode.nodeType === Node.ELEMENT_NODE`.

craigpetchell replied:
Ok. So nodejs doesn't know about `Node`, so I'd need to import a library for building - any preferences here?

treshugart replied:
I'm confused as to why this would be being evaluated in node. Maybe just leave it as is for now. We can change it later.

treshugart commented on src/index.js:
```diff
@@ -609,12 +610,40 @@ const members = {
       eachChildNode(this, node => {
-        const getOuterHtml = getOuterHtmlByNodeType[node.nodeType] || getHtmlNodeOuterHtml;
+        let getOuterHtml;
+        if (node.nodeType === 3) {
```
`3` -> `Node.TEXT_NODE`

craigpetchell replied:
That'll teach me for following the prior art established in `getOuterHtmlByNodeType`. Will fix. 🐼

treshugart replied:
Yeah, we should have done this from the start. Sorry :)

Commit:
style: Use constants for node type from Node instead of magic numbers.

Commit:
refactor: Refactor and optimise check if parent node is a node type that should rendering innerHTML

Commit:
Merge remote-tracking branch 'origin/master' into fix-innerhtml-escaping

Commit:
perf: Optimise rawNode check as don't need to walk up tree.

Walking up the tree was really only necessary for noscript and that default behaviour is already

inconsistent between FF and Chrome.

craigpetchell commented on src/index.js:
```diff
@@ -607,14 +608,43 @@ const members = {
-        const getOuterHtml = getOuterHtmlByNodeType[node.nodeType] || getHtmlNodeOuterHtml;
+        let getOuterHtml;
+        if (node.nodeType === Node.TEXT_NODE) {
+          if (isCommonNodeRawText || isRawTextNode(node)) {
```
(derp) isRawTextNode(node) makes no sense here (as it's not an element).

Commit:
fix: Remove checking if text node is a raw text element (redundant)

craigpetchell commented:
@treshugart @nadiam84 anything left to follow up here (other than a green build - rerunning as it just seemed to time out).

treshugart commented:
lgtm

adevnadia commented:
lgtm
